dev_tools/scripts/generate_cuda_memory_ptx.py

```diff
@@ -186,14 +181,109 @@ class type_desc:
 {{
     auto real_result = result.real();
     auto imag_result = result.imag();
+    asm volatile(
+#if __CUDA_ARCH__ < 700
+        "st.volatile{s.ptx_space_suffix}.v2{t.ptx_type_suffix} [%0], {{%1, %2}};"
+#else
+        "st.relaxed{s.ptx_scope_suffix}{s.ptx_space_suffix}.v2{t.ptx_type_suffix} [%0], {{%1, %2}};"
+#endif
+        :: "{s.ptr_constraint}"({mut_ptr_expr}), "{t.val_constraint}"(real_result), "{t.val_constraint}"(imag_result)
+        : "memory");
+}}
+""")
+
+# since there are no constraints for f16 register an intermediate conversion needs to happen
+t = type_desc(ptx_type_suffix='.f16', val_constraint='f', name='__half')
+t.parent_name = "float"
+t.ptx_parent_type_suffix = '.f32'
+t.ptx_mem_type_suffix = '.b16'
+for s in memory_spaces:
+    for o in memory_orderings:
+        membar_expression = "" if o.is_relaxed else f"membar_acq_rel{s.fn_suffix}();"
+        const_ptr_expr = s.ptr_expr.format(
+            ptr=f"const_cast<{t.name}*>(ptr)")
+        mut_ptr_expr = s.ptr_expr.format(ptr="ptr")
+        print(f"""
+__device__ __forceinline__ {t.name} load{o.fn_load_suffix}{s.fn_suffix}(const {t.name}* ptr)
+{{
+    {t.parent_name} result;
+    asm volatile("{{\\n\\t"
```

Comment:
This complexity makes me ask: Do we need 16 bit precision triangular solves? IIRC the results from Fritz were not showing notable speedups.

Reply:
Yes, we got improvement in Petr's example

Reply:
but what we need for half precision is only 2-4 functions (no need for acqrel, and in shared memory it's likely more efficient to use float), can we just write them out explicitly?